File: genshin-craft-paper/src/main/java/ltd/kumo/plutomc/framework/shared/command/Command.java
package ltd.kumo.plutomc.framework.shared.command;

import ltd.kumo.plutomc.framework.shared.command.executors.Executor;
import ltd.kumo.plutomc.framework.shared.command.executors.PlayerExecutor;
import org.bukkit.command.CommandSender;

import java.util.function.Consumer;
import java.util.function.Predicate;

public interface Command {

    /**
     * 获取本指令的名称
     *
     * @return 名称
     */
    String name();

    Command suggests(Consumer<Suggestion> provider);

    /**
     * 判断一个发送者是否有权限执行，如果返回值是false，则不会提供对应的指令提示
     *
     * @param requirement 需求
     * @return 自己
     */
    Command requires(Predicate<CommandSender> requirement);

    /**
     * 为指令添加别名
     *
     * @param aliases 别名
     * @return 自己
     */
    Command aliases(String... aliases);

    /**
     * 设置执行器
     *
     * @param executor 执行器
     * @return 自己
     */
    Command executes(Executor executor);

    /**
     * 设置玩家执行器，不是玩家执行不会调用执行器
     *
     * @param executor 执行器
     * @return 自己
     */
    Command executesPlayer(PlayerExecutor executor);

    /**
     * 添加指令子节点
     *
     * @param name 指令名称
     * @return 子节点指令
     */
    Command then(String name);

    /**
     * 添加参数子节点
     *
     * @param name 参数名称
     * @param type 参数类型的类
     * @param <E>  参数类型
     * @return 子节点指令
     */
    <E extends Argument<?>> Command then(String name, Class<E> type);

    /**
     * 添加整型参数子节点
     *
     * @param name 参数名称
     * @param min  最小值
     * @param max  最大值
     * @return 子节点指令
     */
    Command thenInteger(String name, int min, int max);

    /**
     * 添加长整型参数子节点
     *
     * @param name 参数名称
     * @param min  最小值
     * @param max  最大值
     * @return 子节点指令
     */
    Command thenLong(String name, long min, long max);

    /**
     * 添加单精度浮点型参数子节点
     *
     * @param name 参数名称
     * @param min  最小值
     * @param max  最大值
     * @return 子节点指令
     */
    Command thenFloat(String name, float min, float max);

    /**
     * 添加双精度浮点型参数子节点
     *
     * @param name 参数名称
     * @param min  最小值
     * @param max  最大值
     * @return 子节点指令
     */
    Command thenDouble(String name, double min, double max);

    /**
     * 复制一个除了名字以外完全相同的指令
     *
     * @param name 新名称
     * @return 新指令
     */
    Command clone(String name);

}

File: genshin-craft-paper/src/main/kotlin/net/deechael/genshincraft/command/Command.kt
package net.deechael.genshincraft.command

import ltd.kumo.plutomc.framework.bukkit.command.BukkitCommand
import ltd.kumo.plutomc.framework.shared.command.Argument
import ltd.kumo.plutomc.framework.shared.command.CommandContext
import net.deechael.genshincraft.GenshinCraft
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class Command {

    private val command: BukkitCommand

    internal constructor(name: String) {
        this.command = GenshinCraft.commandManager.createCommand(name)
    }

    internal constructor(command: BukkitCommand) {
        this.command = command
    }

    fun original(): BukkitCommand {
        return this.command
    }
}

@Target(AnnotationTarget.VALUE_PARAMETER, AnnotationTarget.FUNCTION, AnnotationTarget.TYPE)
annotation class CommandBody {

}

@CommandBody
fun Literal(name: String, body: @CommandBody Command.() -> Unit): Command {
    val command = Command(name)
    command.apply(body)
    return command
}

@CommandBody
fun Command.Literal(name: String, body: @CommandBody Command.() -> Unit): Command {
    val command = Command(name)
    command.apply(body)
    this.original().then(command.original())
    return command
}

@CommandBody
fun Command.Argument(name: String, type: Class<out Argument<*>>, body: @CommandBody Command.() -> Unit): Command {
    val command = Command(this.original().then(name, type))
    command.apply(body)
    this.original().then(command.original())
    return command
}

@CommandBody
fun Command.Executor(body: @CommandBody (CommandSender, CommandContext) -> Unit) {
    this.original().apply {
        this.executes(body)
    }
}

@CommandBody
fun Command.PlayerExecutor(body: @CommandBody (Player, CommandContext) -> Unit) {
    this.original().apply {
        this.executesPlayer(body)
    }
}

@CommandBody
fun Command.Requirement(body: @CommandBody Command.(CommandSender) -> Boolean) {
    this.original().apply {
        this.requires {
            body(it)
        }
    }
}

@CommandBody
fun Command.TextSuggestion(vararg suggestions: String) {
    this.original().apply {
        this.suggests { suggestion ->
            suggestions.forEach {
                suggestion.suggests(it)
            }
        }
    }
}

@CommandBody
fun Command.IntSuggestion(vararg suggestions: Int) {
    this.original().apply {
        this.suggests { suggestion ->
            suggestions.forEach {
                suggestion.suggests(it)
            }
        }
    }
}
